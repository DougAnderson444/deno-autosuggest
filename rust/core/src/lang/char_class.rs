use std::fmt;
use super::Lang;


pub trait CharPattern: fmt::Debug {
    fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool>;
}


#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum CharClass {
    Any,
    Control,
    Whitespace,
    Punctuation,
    NotAlphaNum,
    Consonant,
    Vowel,
}

use CharClass::{
    Any,
    Control,
    Whitespace,
    Punctuation,
    NotAlphaNum,
    Consonant,
    Vowel,
};


impl CharPattern for CharClass {
    fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> {
        match self {
            Any         => Some(true),
            Control     => Some(ch.is_control()),
            Whitespace  => Some(ch.is_whitespace()),
            Punctuation => Some(ch.is_ascii_punctuation()),
            NotAlphaNum => Some(!ch.is_alphanumeric()),
            Consonant   => Some(lang.as_ref()?.get_char_class(ch)? == Consonant),
            Vowel       => Some(lang.as_ref()?.get_char_class(ch)? == Vowel),
        }
    }
}


impl<P: CharPattern> CharPattern for [P] {
    fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> {
        let mut met_none = false;
        for pattern in self {
            match pattern.matches(ch, lang) {
                Some(true)  => return Some(true),
                Some(false) => continue,
                None => {
                    met_none = true;
                    continue;
                },
            }
        }
        if met_none { None } else { Some(false) }
    }
}


impl<P: CharPattern> CharPattern for [P; 1] { fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> { self[..].matches(ch, lang) } }
impl<P: CharPattern> CharPattern for [P; 2] { fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> { self[..].matches(ch, lang) } }
impl<P: CharPattern> CharPattern for [P; 3] { fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> { self[..].matches(ch, lang) } }
impl<P: CharPattern> CharPattern for [P; 4] { fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> { self[..].matches(ch, lang) } }
impl<P: CharPattern> CharPattern for [P; 5] { fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> { self[..].matches(ch, lang) } }


#[cfg(test)]
mod tests {
    use super::{CharClass, CharPattern, Lang};

    use CharClass::{
        Any,
        Control,
        Whitespace,
        Punctuation,
        NotAlphaNum,
        Consonant,
        Vowel,
    };

    #[test]
    fn pattern_matches_any() {
        assert_eq!(Any.matches('\0', &None), Some(true));
        assert_eq!(Any.matches('2',  &None), Some(true));
        assert_eq!(Any.matches('f',  &None), Some(true));
        assert_eq!(Any.matches(';',  &None), Some(true));
        assert_eq!(Any.matches(' ',  &None), Some(true));
    }

    #[test]
    fn pattern_matches_control() {
        assert_eq!(Control.matches('\0', &None), Some(true));
        assert_eq!(Control.matches('2',  &None), Some(false));
        assert_eq!(Control.matches('f',  &None), Some(false));
        assert_eq!(Control.matches(';',  &None), Some(false));
        assert_eq!(Control.matches(' ',  &None), Some(false));
    }

    #[test]
    fn pattern_matches_whitespaces() {
        assert_eq!(Whitespace.matches(' ',  &None), Some(true));
        assert_eq!(Whitespace.matches('\t', &None), Some(true));
        assert_eq!(Whitespace.matches('\n', &None), Some(true));
        assert_eq!(Whitespace.matches('2',  &None), Some(false));
        assert_eq!(Whitespace.matches('f',  &None), Some(false));
        assert_eq!(Whitespace.matches(';',  &None), Some(false));
        assert_eq!(Whitespace.matches('\0', &None), Some(false));
    }

    #[test]
    fn pattern_matches_punctuation() {
        assert_eq!(Punctuation.matches(';',  &None), Some(true));
        assert_eq!(Punctuation.matches('.',  &None), Some(true));
        assert_eq!(Punctuation.matches(',',  &None), Some(true));
        assert_eq!(Punctuation.matches('2',  &None), Some(false));
        assert_eq!(Punctuation.matches('f',  &None), Some(false));
        assert_eq!(Punctuation.matches(' ',  &None), Some(false));
        assert_eq!(Punctuation.matches('\0', &None), Some(false));
    }

    #[test]
    fn pattern_matches_non_alphanum() {
        assert_eq!(NotAlphaNum.matches('2',  &None), Some(false));
        assert_eq!(NotAlphaNum.matches('f',  &None), Some(false));
        assert_eq!(NotAlphaNum.matches('ы',  &None), Some(false));
        assert_eq!(NotAlphaNum.matches('も', &None), Some(false));
        assert_eq!(NotAlphaNum.matches(';',  &None), Some(true));
        assert_eq!(NotAlphaNum.matches('.',  &None), Some(true));
        assert_eq!(NotAlphaNum.matches(',',  &None), Some(true));
        assert_eq!(NotAlphaNum.matches(' ',  &None), Some(true));
        assert_eq!(NotAlphaNum.matches('\0', &None), Some(true));
    }

    #[test]
    fn pattern_matches_slice() {
        let pattern = &[Whitespace, Punctuation][..];

        assert_eq!(pattern.matches(' ',  &None), Some(true));
        assert_eq!(pattern.matches(';',  &None), Some(true));
        assert_eq!(pattern.matches('f',  &None), Some(false));
        assert_eq!(pattern.matches('2',  &None), Some(false));
        assert_eq!(pattern.matches('\0', &None), Some(false));
    }

    #[test]
    fn pattern_matches_array_2() {
        let pattern = [Whitespace, Punctuation];

        assert_eq!(pattern.matches(' ',  &None), Some(true));
        assert_eq!(pattern.matches(';',  &None), Some(true));
        assert_eq!(pattern.matches('f',  &None), Some(false));
        assert_eq!(pattern.matches('2',  &None), Some(false));
        assert_eq!(pattern.matches('\0', &None), Some(false));
    }

    #[test]
    fn pattern_matches_vowel_no_lang() {
        assert_eq!(Vowel.matches('2',  &None), None);
        assert_eq!(Vowel.matches('f',  &None), None);
        assert_eq!(Vowel.matches('ы',  &None), None);
        assert_eq!(Vowel.matches('も', &None), None);
        assert_eq!(Vowel.matches(';',  &None), None);
        assert_eq!(Vowel.matches('.',  &None), None);
        assert_eq!(Vowel.matches(',',  &None), None);
        assert_eq!(Vowel.matches(' ',  &None), None);
        assert_eq!(Vowel.matches('\0', &None), None);
    }

    #[test]
    fn pattern_matches_vowel_with_lang() {
        let mut lang = Lang::new();
        lang.add_char_class('f', Consonant);
        lang.add_char_class('ы', Vowel);
        let lang = Some(lang);

        assert_eq!(Vowel.matches('2',  &lang), None);
        assert_eq!(Vowel.matches('f',  &lang), Some(false));
        assert_eq!(Vowel.matches('ы',  &lang), Some(true));
        assert_eq!(Vowel.matches('も', &lang), None);
        assert_eq!(Vowel.matches(';',  &lang), None);
        assert_eq!(Vowel.matches('.',  &lang), None);
        assert_eq!(Vowel.matches(',',  &lang), None);
        assert_eq!(Vowel.matches(' ',  &lang), None);
        assert_eq!(Vowel.matches('\0', &lang), None);
    }

    #[test]
    fn pattern_matches_consonant_no_lang() {
        assert_eq!(Consonant.matches('2',  &None), None);
        assert_eq!(Consonant.matches('f',  &None), None);
        assert_eq!(Consonant.matches('ы',  &None), None);
        assert_eq!(Consonant.matches('も', &None), None);
        assert_eq!(Consonant.matches(';',  &None), None);
        assert_eq!(Consonant.matches('.',  &None), None);
        assert_eq!(Consonant.matches(',',  &None), None);
        assert_eq!(Consonant.matches(' ',  &None), None);
        assert_eq!(Consonant.matches('\0', &None), None);
    }

    #[test]
    fn pattern_matches_consonant_with_lang() {
        let mut lang = Lang::new();
        lang.add_char_class('f', Consonant);
        lang.add_char_class('ы', Vowel);
        let lang = Some(lang);

        assert_eq!(Consonant.matches('2',  &lang), None);
        assert_eq!(Consonant.matches('f',  &lang), Some(true));
        assert_eq!(Consonant.matches('ы',  &lang), Some(false));
        assert_eq!(Consonant.matches('も', &lang), None);
        assert_eq!(Consonant.matches(';',  &lang), None);
        assert_eq!(Consonant.matches('.',  &lang), None);
        assert_eq!(Consonant.matches(',',  &lang), None);
        assert_eq!(Consonant.matches(' ',  &lang), None);
        assert_eq!(Consonant.matches('\0', &lang), None);
    }
}
