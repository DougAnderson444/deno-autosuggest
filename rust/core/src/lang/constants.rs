use super::CharClass;
use CharClass::{Consonant, Vowel};

pub const CHAR_CLASSES_LATIN: &[(CharClass, char)] = &[
    (Consonant, 'b'),
    (Consonant, 'c'),
    (Consonant, 'd'),
    (Consonant, 'f'),
    (Consonant, 'g'),
    (Consonant, 'h'),
    (Consonant, 'j'),
    (Consonant, 'k'),
    (Consonant, 'l'),
    (Consonant, 'm'),
    (Consonant, 'n'),
    (Consonant, 'p'),
    (Consonant, 'q'),
    (Consonant, 'r'),
    (Consonant, 's'),
    (Consonant, 't'),
    (Consonant, 'v'),
    (Consonant, 'w'),
    (Consonant, 'x'),
    (Consonant, 'z'),
    (Vowel, 'a'),
    (Vowel, 'e'),
    (Vowel, 'i'),
    (Vowel, 'o'),
    (Vowel, 'u'),
    (Vowel, 'y'),
];
