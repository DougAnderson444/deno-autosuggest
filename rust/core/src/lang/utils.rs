use std::collections::HashMap;
use crate::utils::to_vec;


pub fn compile_utf_map(static_map: &[(&str, &str)]) -> HashMap<Vec<char>, Vec<char>> {
    let mut compiled_map = HashMap::with_capacity(static_map.len());
    for (pattern, replace) in static_map {
        compiled_map.insert(to_vec(pattern), to_vec(replace));
    }
    compiled_map
}
