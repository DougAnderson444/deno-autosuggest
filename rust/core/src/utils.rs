#![macro_use]

use std::cell::RefCell;


macro_rules! min {
    ($x: expr) => ($x);
    ($x: expr, $($z: expr),+) => (::std::cmp::min($x, min!($($z),*)));
}


macro_rules! max {
    ($x: expr) => ($x);
    ($x: expr, $($z: expr),+) => (::std::cmp::max($x, max!($($z),*)));
}
