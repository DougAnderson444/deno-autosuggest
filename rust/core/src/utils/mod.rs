#![macro_use]

mod trigrams;
mod limitsort;
mod fading_windows;

pub use trigrams::{Trigrams, TrigramIter};
pub use limitsort::{LimitSort, LimitSortIter};
pub use fading_windows::FadingWindows;


macro_rules! min {
    ($x: expr) => ($x);
    ($x: expr, $($z: expr),+) => (::std::cmp::min($x, min!($($z),*)));
}


macro_rules! max {
    ($x: expr) => ($x);
    ($x: expr, $($z: expr),+) => (::std::cmp::max($x, max!($($z),*)));
}


pub fn to_vec<T: AsRef<str>>(chars: T) -> Vec<char> {
    chars.as_ref().chars().collect()
}


#[cfg(test)]
pub fn to_str<T: AsRef<[char]>>(chars: T) -> String {
    chars.as_ref().iter().collect()
}
