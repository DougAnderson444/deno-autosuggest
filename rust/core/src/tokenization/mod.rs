mod word;
mod text;

use crate::lang::{Lang, Chars, CharPattern};
pub use word::Word;
pub use text::Text;


pub fn tokenize_query(source: &str, lang: &Option<Lang>) -> Text<Vec<char>> {
    Text::from_str(source)
        .normalize(lang)
        .fin(false)
        .split(&[Chars::Whitespaces, Chars::Control, Chars::Punctuation])
        .strip(&[Chars::NotAlphaNum])
        .lower()
        .mark_pos(lang)
        .stem(lang)
}


pub fn tokenize_record(source: &str, lang: &Option<Lang>) -> Text<Vec<char>> {
     Text::from_str(source)
        .normalize(lang)
        .split(&[Chars::Whitespaces, Chars::Control, Chars::Punctuation])
        .strip(&[Chars::NotAlphaNum])
        .lower()
        .mark_pos(lang)
        .stem(lang)
}
