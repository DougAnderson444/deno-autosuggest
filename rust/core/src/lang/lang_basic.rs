#![allow(dead_code)]

use super::Lang;
use super::constants::CHAR_CLASSES_LATIN;

pub fn lang_basic() -> Lang {
    let mut lang = Lang::new();
    for &(class, ch) in CHAR_CLASSES_LATIN { lang.add_char_class(ch, class); }
    lang
}
