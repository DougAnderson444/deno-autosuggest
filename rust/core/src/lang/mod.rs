mod normalize;
mod lang_english;
mod lang_german;
mod lang_portuguese;
mod lang_russian;
mod lang_spanish;

pub use lang_german::lang_german;
pub use lang_english::lang_english;
pub use lang_spanish::lang_spanish;
pub use lang_portuguese::lang_portuguese;
pub use lang_russian::lang_russian;

use std::cell::RefCell;
use fnv::{FnvHashMap as HashMap};
use rust_stemmers::Stemmer;
use crate::utils::to_vec;
use crate::tokenization::PartOfSpeech;
use normalize::Normalize;


const BUFFER_CAPACITY: usize = 20;


pub struct Lang {
    stemmer:      Option<Stemmer>,
    pos_map:      HashMap<Vec<char>, PartOfSpeech>,
    compose_map:  HashMap<Vec<char>, Vec<char>>,
    reduce_map:   HashMap<Vec<char>, Vec<char>>,
    stem_buffer:  RefCell<String>,
    norm_buffer1: RefCell<Vec<char>>,
    norm_buffer2: RefCell<Vec<char>>,
}


impl Lang {
    pub fn new() -> Self {
        Self {
            stemmer:      None,
            pos_map:      HashMap::default(),
            compose_map:  HashMap::default(),
            reduce_map:   HashMap::default(),
            stem_buffer:  RefCell::new(String::with_capacity(BUFFER_CAPACITY)),
            norm_buffer1: RefCell::new(Vec::with_capacity(BUFFER_CAPACITY)),
            norm_buffer2: RefCell::new(Vec::with_capacity(BUFFER_CAPACITY)),
        }
    }

    pub fn set_stemmer(&mut self, stemmer: Option<Stemmer>) {
        self.stemmer = stemmer;
    }

    pub fn add_pos(&mut self, word: &str, pos: PartOfSpeech) {
        let source   = to_vec(word);
        let composed = self.unicode_compose(&source).unwrap_or(source);
        let reduced  = self.unicode_reduce(&composed);

        self.pos_map.insert(composed, pos);

        if let Some((_, reduced)) = reduced {
            self.pos_map.insert(reduced, pos);
        }
    }

    pub fn add_unicode_composition(&mut self, from: &str, to: &str) {
        self.compose_map.insert(to_vec(from), to_vec(to));
    }

    pub fn add_unicode_reduction(&mut self, from: &str, to: &str) {
        self.reduce_map.insert(to_vec(from), to_vec(to));
    }

    pub fn stem(&self, word: &[char]) -> usize {
        if let Some(stemmer) = &self.stemmer {
            let buffer = &mut *self.stem_buffer.borrow_mut();
            buffer.clear();
            buffer.extend(word.iter());
            let stem = stemmer.stem(&buffer);
            stem.chars().count()
        } else {
            word.len()
        }
    }

    pub fn get_pos(&self, word: &[char]) -> Option<PartOfSpeech> {
        self.pos_map.get(word).cloned()
    }

    pub fn unicode_compose(&self, word: &[char]) -> Option<Vec<char>> {
        let buffer = &mut *self.norm_buffer1.borrow_mut();
        buffer.clear();

        for (_, norm_chunk) in Normalize::new(word, &self.compose_map) {
            buffer.extend(norm_chunk);
        }

        if &buffer[..] == word {
            None
        } else {
            Some(buffer.clone())
        }
    }

    pub fn unicode_reduce(&self, word: &[char]) -> Option<(Vec<char>, Vec<char>)> {
        let buffer1 = &mut *self.norm_buffer1.borrow_mut();
        let buffer2 = &mut *self.norm_buffer2.borrow_mut();
        buffer1.clear();
        buffer2.clear();

        for (word_chunk, norm_chunk) in Normalize::new(word, &self.reduce_map) {
            buffer1.extend(word_chunk);
            buffer2.extend(norm_chunk);
            for _ in 0 .. norm_chunk.len() - word_chunk.len() {
                buffer1.push('\0');
            }
        }

        if &buffer2[..] == word {
            None
        } else {
            Some((buffer1.clone(), buffer2.clone()))
        }
    }
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::utils::to_vec;
    use crate::tokenization::PartOfSpeech;
    use super::Lang;

    fn get_lang() -> Lang {
        let mut lang = Lang::new();
        lang.add_unicode_composition("ó", "ó");
        lang.add_unicode_reduction("ó", "o");
        lang.add_unicode_reduction("õ", "oo");
        lang.add_pos("fóo", PartOfSpeech::Particle);
        lang
    }

    #[test]
    fn unicode_compose_nfc() {
        let input  = to_vec("foóbar");
        let output = get_lang().unicode_compose(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn unicode_compose_nfd() {
        let input  = to_vec("foóbar");
        let output = get_lang().unicode_compose(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn unicode_reduce_reduced() {
        let input  = to_vec("foobar");
        let output = get_lang().unicode_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn unicode_reduce_nfc() {
        let input  = to_vec("foóbar");
        let output = get_lang().unicode_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn unicode_reduce_fill0() {
        let input  = to_vec("fõbar");
        let output = get_lang().unicode_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn get_pos_source() {
        let input = to_vec("fóo");
        let pos   = get_lang().get_pos(&input);
        assert_eq!(pos, Some(PartOfSpeech::Particle));
    }

    #[test]
    fn get_pos_reduced() {
        let input = to_vec("foo");
        let pos   = get_lang().get_pos(&input);
        assert_eq!(pos, Some(PartOfSpeech::Particle));
    }

    #[test]
    fn get_pos_unknown() {
        let input = to_vec("bar");
        let pos   = get_lang().get_pos(&input);
        assert_eq!(pos, None);
    }
}