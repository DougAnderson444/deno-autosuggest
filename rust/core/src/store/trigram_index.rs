use std::cmp::Ordering;
use fnv::{FnvHashMap as HashMap};
use crate::utils::{Trigrams, LimitSort};
use crate::tokenization::{Word, TextRef};
use super::Record;

use Ordering::{
    Less,
    Equal,
    Greater,
};


pub struct TrigramIndex {
    len:     usize,
    dict:    HashMap<[char; 3], Vec<usize>>,
    counts:  Vec<usize>,
}


impl TrigramIndex {
    pub fn new() -> Self {
        Self {
            len:    0,
            dict:   HashMap::default(),
            counts: Vec::new(),
        }
    }

    pub fn add(&mut self, record: &Record) {
        let Self { dict, len, .. } = self;
        let Record { ix, title, .. } = record;
        let grams = Self::collect_grams(&title.to_ref());
        *len += 1;
        for &gram in grams.iter() {
            dict
                .entry(gram)
                .and_modify(|ixs| {
                    for i in 0..ixs.len() {
                        match ix.cmp(unsafe { ixs.get_unchecked(i) }) {
                            Greater => continue,
                            Equal   => return,
                            Less    => return ixs.insert(i, *ix),
                        }
                    }
                    ixs.push(*ix);
                })
                .or_insert_with(|| vec![*ix]);
        }
    }

    pub fn prepare(
        &mut self,
        query:   &TextRef,
        size:    usize,
    ) -> Vec<usize> {
        let Self { counts, dict, .. } = self;

        if query.words.len() == 0 {
            return Vec::new();
        }

        counts.clear();
        counts.resize(self.len, 0);

        let grams = Self::collect_grams(&query);

        for gram in grams.iter() {
            if let Some(ixs) = dict.get(gram) {
                for &ix in ixs {
                    unsafe {
                        *counts.get_unchecked_mut(ix) += 1;
                    }
                }
            }
        }

        counts
            .iter()
            .enumerate()
            .filter(|(_, &count)| count > 0)
            .limit_sort_unstable(size * 10, |(_, count1), (_, count2)| count2.cmp(count1))
            .map(|(ix, _)| ix)
            .collect()
    }

    fn collect_grams(text: &TextRef) -> Vec<[char; 3]> {
        let cap       = text.words.iter().map(|w| w.len()).sum::<usize>();
        let mut grams = Vec::with_capacity(cap);
        for word in text.words {
            let chars = &text.chars[word.slice.0 .. word.slice.1];
            for gram in chars.trigrams() {
                grams.push(gram);
            }
        }
        grams.sort_unstable();
        grams.dedup();
        grams
    }
}


#[cfg(test)]
mod tests {
    use insta::{assert_debug_snapshot, assert_snapshot};
    use crate::lang::Lang;
    use crate::tokenization::tokenize_query;
    use super::Record;
    use super::TrigramIndex;

    fn get_index() -> (TrigramIndex, [Record; 5]) {
        let lang = Lang::new();
        let mut records = [
            Record::new(10, "brown plush bear",     10, &lang),
            Record::new(20, "the metal detector",   20, &lang),
            Record::new(30, "yellow metal mailbox", 30, &lang),
            Record::new(40, "thesaurus",            40, &lang),
            Record::new(50, "wi-fi router",         50, &lang),
        ];
        let mut index = TrigramIndex::new();
        for (ix, record) in records.iter_mut().enumerate() {
            record.ix = ix;
            index.add(&record);
        }
        (index, records)
    }

    fn export_dict(index: &TrigramIndex) -> String {
        let mut dict = index.dict
            .iter()
            .map(|(gram, ixs)| (
                gram.iter().cloned().collect::<String>(),
                ixs.clone(),
            ))
            .collect::<Vec<_>>();

        dict.sort_by(|(g1, _), (g2, _)| g1.cmp(g2));

        let mut result = String::new();
        for (gram, ixs) in &dict {
            result.push_str(&format!("\"{:3}\" {:?}\n", gram, ixs));
        }
        result
    }

    fn check_prepare(name: &str, size: usize, queries: &[&str]) {
        let lang = Lang::new();
        let (mut index, _) = get_index();
        for (i, query) in queries.iter().enumerate() {
            let query = tokenize_query(query, &lang);
            let query = query.to_ref();
            let mut prepared = index.prepare(&query, size);
            dbg!(&query);
            prepared.sort();
            assert_debug_snapshot!(format!("{}-{}", name, i), prepared);
        }
    }

    #[test]
    fn add_first() {
        let lang       = Lang::new();
        let mut index  = TrigramIndex::new();
        index.add(&Record::new(10, "Foo Bar", 10, &lang));
        assert_snapshot!(export_dict(&index));
    }

    #[test]
    fn add_second() {
        let lang        = Lang::new();
        let mut index   = TrigramIndex::new();
        let mut record1 = Record::new(10, "Foo Bar", 10, &lang);
        let mut record2 = Record::new(20, "Bar Baz", 20, &lang);
        record1.ix = 0;
        record2.ix = 1;
        index.add(&record1);
        index.add(&record2);
        assert_snapshot!(export_dict(&index));
    }

    #[test]
    fn prepare_mismatch() {
        check_prepare("mismatch", 3, &["zzzap!"]);
    }

    #[test]
    fn prepare_all_matches() {
        check_prepare("all_matches", 3, &[
            "metal",
            "the",
            "rou",
        ]);
    }

    #[test]
    fn prepare_empty() {
        check_prepare("empty", 3, &[
            "",
        ]);
    }

    #[test]
    fn prepare_first_char() {
        check_prepare("first_char", 3, &[
            "m",
            "t",
            "r",
        ]);
    }

    #[test]
    fn prepare_two_chars() {
        check_prepare("two_chars", 3, &[
            "me",
            "th",
        ]);
    }
}
