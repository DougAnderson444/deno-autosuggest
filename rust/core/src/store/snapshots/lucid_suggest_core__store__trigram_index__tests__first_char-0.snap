---
source: src/store/trigram_index.rs
expression: prepared
---
[
    1,
    2,
]
